### DISCUSSION TITLE
Add an include profile for dbus system tray access

### DISCUSSION THREAD

**Discussion by kmk3** [netblue30/firejail#4524] [Ideas]
(Continuing the discussion from #4510)

Note: I don't know much about these filters and I'm not sure if it's worth it
to add a profile just for 2 commands, so currently this is more of a
hypothetical scenario and out of curiosity about why not "allow-\*.inc includes
for D-Bus".

@kmk3 commented [on Sep 9](https://github.com/netblue30/firejail/pull/4510#issuecomment-916450763):

> > ```diff
> > diff --git a/etc/firejail.config b/etc/firejail.config
> > index 2e355586b..5111bb769 100644
> > --- a/etc/firejail.config
> > +++ b/etc/firejail.config
> > @@ -2,6 +2,9 @@
> >  # keyword-argument pairs, one per line. Most features are enabled by default.
> >  # Use 'yes' or 'no' as configuration values.
> >  
> > +# Allow programs to display a tray icon
> > +# allow-tray no
> > +
> >  # Enable AppArmor functionality, default enabled.
> >  # apparmor yes
> > ```

[...]

> So is this intended to just allow interacting with the system tray through
> dbus? If so, wouldn't it suffice to add some dbus filters like the above in a
> new allow-system-tray.inc file?

@rusty-snake commented [on Sep 10](https://github.com/netblue30/firejail/pull/4510#issuecomment-916657799):

> > So is this intended to just allow interacting with the system tray through
> > dbus?
> 
> Yes
> 
> > If so, wouldn't it suffice to add some dbus filters like the above in a new
> > allow-system-tray.inc file?
> 
> It's a different idea. For tray¹ icons to work you need `dbus-user.talk
> org.kde.StatusNotifierWatcher` and, depending on the implementation,
> `dbus-user.own org.kde.*`. Both are unsafe and you don't want them allowed by
> default. On the other hand some users might heavily use tray icons. Therefore
> the Idea is to have a option in firejail.config which enables tray icons
> globally (profiles require `?ALLOW_TRAY: dbus-user.talk
> org.kde.StatusNotifierWatcher`) instead of creating a lot .locals. This is
> much easier for novice users, especially if it is integrated into
> firejail-welcome.sh.
> 
> ¹ There are many tray icon standards but this is the most common one.
> 
> General: I have already argued against allow-\*.inc includes for D-Bus.

Why is that?  Is there a significant difference compared to filesystem paths?
What if instead of dbus filters, the mechanism used was e.g.: sockets?
Example:

```firejail
noblacklist ${RUNUSER}/foo/systemtray
noblacklist ${RUNUSER}/kde/statusnotifier
noblacklist ${RUNUSER}/xdg/statusnotifier

whitelist ${RUNUSER}/foo/systemtray
whitelist ${RUNUSER}/kde/statusnotifier
whitelist ${RUNUSER}/xdg/statusnotifier
```

I don't know if this exists, but for the sake of argument, what if a program
used dbus by default and something like the above as a fallback?

Anyway, I think that an allow- profile could be complementary to `ALLOW_TRAY`,
especially considering that:

* There are multiple filters intended to do the same thing
* I'd guess that more may be more added eventually
* A given program may depend on any combination of them

Without the include, the following (or parts of it) has to be copied to every
relevant profile:

```firejail
dbus-user.talk foo.bar
dbus-user.talk org.freedesktop.StatusNotifierItem
dbus-user.talk org.kde.StatusNotifierWatcher
```

With only `ALLOW_TRAY`:

```firejail
?ALLOW_TRAY: dbus-user.talk foo.bar
?ALLOW_TRAY: dbus-user.talk org.freedesktop.StatusNotifierItem
?ALLOW_TRAY: dbus-user.talk org.kde.StatusNotifierWatcher
```

With `ALLOW_TRAY` and the allow- profile:

```firejail
?ALLOW_TRAY: include allow-dbus-system-tray.inc
```

**Comment by rusty-snake** [comment]
> dbus-user.talk org.freedesktop.StatusNotifierItem

- I never saw a app using the freedesktop spec, all app I saw used the kde spec.
  - The freeesktop spec it the kde spec with changed names.
- `org.kde.StatusNotifierItem` is the interface every app needs provide.
  - To provide, it needs to be an `own` rule.
  - Apps using libayatana (common for GTK programs) need to `own` their bus name (`com.example.App`)
  - However the spec says `org.kde.StatusNotifierItem-PID-ID` and KDE/Qt libs follow this
  - To express this for xdg-dbus-proxy, you need to `own org.kde.*`

> With only ALLOW_TRAY:
>
> ```
> ?ALLOW_TRAY: dbus-user.talk foo.bar
> ?ALLOW_TRAY: dbus-user.talk org.freedesktop.StatusNotifierItem
> ?ALLOW_TRAY: dbus-user.talk org.kde.StatusNotifierWatcher
> ```

Actually it would be either
```
?ALLOW_TRAY: dbus-user.own com.example.App
?ALLOW_TRAY: dbus-user.talk org.kde.StatusNotifierWatcher
```
or
```
?ALLOW_TRAY: dbus-user.own org.kde.*
?ALLOW_TRAY: dbus-user.talk org.kde.StatusNotifierWatcher
```

> With ALLOW_TRAY and the allow- profile:
> ```
> ?ALLOW_TRAY: include allow-dbus-system-tray.inc
> ```

So what do you want to put into `allow-dbus-system-tray.inc`?
- `com.example.App`: No
- `org.kde.StatusNotifierWatcher`: Yes
- `org.kde.*`: Do we want Apps using `com.example.App` to give this permission?

**Reply by kmk3** [reply]
@rusty-snake [on Sep 11](https://github.com/netblue30/firejail/discussions/4524#discussioncomment-1309507):

> > dbus-user.talk org.freedesktop.StatusNotifierItem
> 
> * I never saw a app using the freedesktop spec, all app I saw used the kde spec.
>   * The freeesktop spec it the kde spec with changed names.

[...]

> Actually it would be either
> 
> ```
> ?ALLOW_TRAY: dbus-user.own com.example.App
> ?ALLOW_TRAY: dbus-user.talk org.kde.StatusNotifierWatcher
> ```
> 
> or
> 
> ```
> ?ALLOW_TRAY: dbus-user.own org.kde.*
> ?ALLOW_TRAY: dbus-user.talk org.kde.StatusNotifierWatcher
> ```
> 
> > With ALLOW_TRAY and the allow- profile:
> > ```
> > ?ALLOW_TRAY: include allow-dbus-system-tray.inc
> > ```
> 
> So what do you want to put into `allow-dbus-system-tray.inc`?
> 
> * `com.example.App`: No
> * `org.kde.StatusNotifierWatcher`: Yes
> * `org.kde.\*`: Do we want Apps using `com.example.App` to give this
>   permission?

Thanks for the details.  I thought that there were more filters involved that
were common between normal profiles.  If it's effectively only one, indeed
there is no point in creating an include profile.

**Comment by rusty-snake** [comment]
>  > General: I have already argued against allow-*.inc includes for D-Bus.
>
>  Why is that? Is there a significant difference compared to filesystem paths?
> What if instead of dbus filters, the mechanism used was e.g.: sockets?

It would be the same. It's not specific for D-Bus, it's against having a dozen includes containing only 2 lines.
IMHO aliases/variables/macros (#3412, #3424) fit this better.
